File: sdk/java/src/main/java/io/kcm/Fact.java
package io.kcm;

public final class Fact {
    public final int subject;
    public final byte predicate;
    public final int object;
    public final double confidence;
    public final byte evidence;
    public final long timestamp;
    public final byte context;
    public final int version;
    public final byte priority;
    public final short owner;

    public Fact(int subject, byte predicate, int object, double confidence,
                byte evidence, long timestamp, byte context, int version,
                byte priority, short owner) {
        if (confidence < 0.0 || confidence > 1.0) {
            throw new IllegalArgumentException("Confidence must be in [0, 1], got " + confidence);
        }
        this.subject = subject;
        this.predicate = predicate;
        this.object = object;
        this.confidence = confidence;
        this.evidence = evidence;
        this.timestamp = timestamp;
        this.context = context;
        this.version = version;
        this.priority = priority;
        this.owner = owner;
    }

    public Fact(int subject, byte predicate, int object, double confidence) {
        this(subject, predicate, object, confidence, (byte) 0, 0L, (byte) 0, 0, (byte) 0, (short) 0);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Fact)) return false;
        Fact f = (Fact) o;
        return subject == f.subject && predicate == f.predicate && object == f.object &&
               Double.compare(confidence, f.confidence) == 0 && evidence == f.evidence &&
               timestamp == f.timestamp && context == f.context && version == f.version &&
               priority == f.priority && owner == f.owner;
    }

    @Override
    public int hashCode() {
        int h = 17;
        h = 31 * h + subject;
        h = 31 * h + predicate;
        h = 31 * h + object;
        h = 31 * h + Double.hashCode(confidence);
        h = 31 * h + evidence;
        h = 31 * h + Long.hashCode(timestamp);
        h = 31 * h + context;
        h = 31 * h + version;
        h = 31 * h + priority;
        h = 31 * h + owner;
        return h;
    }

    @Override
    public String toString() {
        return String.format("Fact{subject=%d, predicate=%d, object=%d, confidence=%.2f, evidence=%d, timestamp=%d, context=%d, version=%d, priority=%d, owner=%d}",
            subject, predicate, object, confidence, evidence, timestamp, context, version, priority, owner);
    }
}

File: sdk/rust/src/fact.rs
use crate::error::SdkError;

#[derive(Debug, Clone)]
pub struct Fact {
    pub subject: u32,
    pub predicate: u8,
    pub object: u32,
    pub confidence: f64,
    pub evidence: u8,
    pub timestamp: i64,
    pub context: u8,
    pub version: i32,
    pub priority: i8,
    pub owner: u16,
}

impl Fact {
    pub fn new(
        subject: u32,
        predicate: u8,
        object: u32,
        confidence: f64,
    ) -> Result<Self, SdkError> {
        if confidence < 0.0 || confidence > 1.0 {
            return Err(SdkError::InvalidArgument(
                "Confidence must be in [0.0, 1.0]".to_string(),
            ));
        }
        if confidence.is_nan() || confidence.is_infinite() {
            return Err(SdkError::InvalidArgument(
                "Confidence must be finite".to_string(),
            ));
        }
        let timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_nanos()
            .min(i64::MAX as u128) as i64;

        Ok(Fact {
            subject,
            predicate,
            object,
            confidence,
            evidence: 0,
            timestamp,
            context: 0,
            version: 1,
            priority: 0,
            owner: 0,
        })
    }

    pub fn with_evidence(mut self, evidence: u8) -> Self {
        self.evidence = evidence;
        self
    }

    pub fn with_context(mut self, context: u8) -> Self {
        self.context = context;
        self
    }

    pub fn with_version(mut self, version: i32) -> Self {
        self.version = version;
        self
    }

    pub fn with_priority(mut self, priority: i8) -> Self {
        self.priority = priority;
        self
    }

    pub fn with_owner(mut self, owner: u16) -> Self {
        self.owner = owner;
        self
    }

    pub fn with_timestamp(mut self, timestamp: i64) -> Self {
        self.timestamp = timestamp;
        self
    }

    pub(crate) fn to_core(&self) -> Result<kcm_core::types::Fact, SdkError> {
        Ok(kcm_core::types::Fact {
            subject: kcm_core::types::SubjectID::new(self.subject),
            predicate: kcm_core::types::PredicateID::new(self.predicate),
            object: kcm_core::types::ObjectID::new(self.object),
            confidence: self.confidence,
            evidence: kcm_core::types::EvidenceID::new(self.evidence),
            timestamp: self.timestamp,
            context: kcm_core::types::ContextID::new(self.context),
            version: self.version,
            priority: self.priority,
            owner: self.owner,
        })
    }

    pub(crate) fn from_core(fact: kcm_core::types::Fact) -> Self {
        Fact {
            subject: fact.subject.0,
            predicate: fact.predicate.0,
            object: fact.object.0,
            confidence: fact.confidence,
            evidence: fact.evidence.0,
            timestamp: fact.timestamp,
            context: fact.context.0,
            version: fact.version,
            priority: fact.priority,
            owner: fact.owner,
        }
    }
}
